File: java/leetcode_grind/Day336.java
package leetcode_grind;

public class Day336 {
    // https://leetcode.com/problems/painting-the-walls/description
    static class Solution {
        public int paintWalls(int[] cost, int[] time) {
            var n = cost.length;
            var dp = new int[n + 1][n + 1];
            for (int i = 1; i < dp.length; i++) {
                dp[n][i] = 1000_000_000;
            }

            for (int i = n - 1; i >= 0; i--) {
                for (int j = 1; j <= n; j++) {
                    var paint = cost[i] + dp[i + 1][Math.max(0, j - time[i] - 1)];
                    var notPaint = dp[i + 1][j];
                    dp[i][j] = Math.min(paint, notPaint);
                }
            }

            return dp[0][n];
        }
    }
}

File: src/leetcode_grind/day_336.rs
// https://leetcode.com/problems/painting-the-walls/description
pub fn paint_walls(cost: Vec<i32>, time: Vec<i32>) -> i32 {
    let n = cost.len();
    let mut dp = vec![vec![0; n + 1]; n + 1];

    for i in 1..=n {
        dp[n][i] = 1000000000;
    }

    for i in (0..n).rev() {
        for remain in 1..=n {
            let paint = cost[i] + dp[i + 1][0.max(remain as i32 - 1 - time[i]) as usize];
            let dont_paint = dp[i + 1][remain];
            dp[i][remain] = paint.min(dont_paint);
        }
    }

    dp[0][n]
}
